### PR TITLE
Cucumber sync hooks

### LINKED ISSUE #4207: [Cucumber + v5] How to make hooks (afterStep) to run and finish before next step is executed?
Hi, 
Using wdio v4 + cucumber i used some afterSteps and know im trying to use the same in v5 + cucumber but it seems not to be running in order.
On my afterStep i have certain validations, i.e. if result.status === failed then i take a screenshot and do some extra steps, the problem is i am able to take the screenshot but im never able to do the extra steps because when the afterStep is executed, the runner is already executing another test case.

This is the code im using on my afterStep for the v5 (updated from v4):
```
afterStep: function (uri, feature, scenario, step, result) {
        if (result.status === 'failed') {
            const path =  './errorShots/'+Date.now()+'.png';
            browser.saveScreenshot(path);
            const allitems = $('.slds-icon-waffle');
            const logoutBut = $('button[title="Logout"]');
            const adminMenu = $('.oneAppLauncherItemList a[title="Parsable Admin"]');
            const sfProfilelogo = $('.profileTrigger');
            const sflogOut = $('.profile-link-label.logout.uiOutputURL');
            this.allitems.waitForDisplayed();
            this.allitems.click();
            this.adminMenu.waitForVisible();
            this.adminMenu.click();
            browser.pause(2000)
            if (this.logoutBut.isVisible()) {   
                this.logoutBut.click();
            }
            this.sfProfilelogo.waitForVisible();
            this.sfProfilelogo.click();
            this.sflogOut.waitForVisible();
            this.sflogOut.click();
        } 
        console.log(step.text + ' is ' + result.status)
    },
```
Sometimes it successfully takes the screenshot of the page with the error, but the other steps are never completed since when it gets there, the runner is already executing another test case..

**WebdriverIO version: v5.11.2
**Mode:Selenium Standalone
**If WDIO Testrunner, running sync/async: sync
**Node.js version: v10.16.0
**NPM version: v6.9.0
**Browser name and version: Chrome
**Platform name and version: MacOS

How can i ensure the afterStep is completely executed before the next step is executed? 
I have tried using promises but im not sure if i was implementing well, any help?

### PR DESCRIPTION
**Pull request by mgrybyk**
## Proposed changes

- ~remove before/after step hooks~, see https://github.com/cucumber/cucumber-js/issues/997
- fix before/after **step** and **scenario** hooks
- somewhat fix before/after **feature** hooks, they work as BeforeAll/AfterAll per worker

fixes #4207

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: packages/wdio-cucumber-framework/src/index.js [hunk 1]
```diff
@@ -149,6 +148,25 @@ class CucumberAdapter {
         mockery.disable()
     }
 
+    addHooks (config) {
```
Comment (mgrybyk) [inline packages/wdio-cucumber-framework/src/index.js]: Make sure that wrapSteps are not wrapping at least webdriverio hooks.

**Comment by mgrybyk** [comment]
If I could somehow determine step type in wrapStep function that would be awesome. 
In such we would be able to add before after step hooks. 
I've noticed that wrap step function also wraps hooks defined by user. I'm pretty sure this is not something we'd like to do, have to take a look if we can handle it somehow. Sorry to say cucumberJs doesn't support it

**Comment by mgrybyk** [comment]
Great news! I've managed to make before/after step hooks work!

File: packages/wdio-sync/src/index.js [hunk 2]
```diff
@@ -15,33 +15,31 @@ const log = logger('@wdio/sync')
...
-
-            /**
-             * no need to modify stack if no stack available
-             */
-            if (!e.stack) {
-                return reject(e)
-            }
+const executeSync = async function (fn, repeatTest = 0, args = []) {
+    try {
+        let res = fn.apply(this, args)
```
Comment (mgrybyk) [inline packages/wdio-sync/src/index.js]: @christian-bromann may you please help a bit with validating change in executeSync? 
In some cases fn.apply returns promise and have to be awaited before proceeded further.
This helps me with running before after step hooks and also fixing issue when wdio/sync and async functions are used together. 

There is an alternative way to handle before after step hooks by wrapping them with future, waiting for it, and then catching error ”cannot run without a fiber” but I don't really like that.

File: packages/wdio-cucumber-framework/src/utils.js [hunk 3]
```diff
@@ -160,3 +167,57 @@ export function getStepFromFeature(feature, pickle, stepIndex, sourceLocation) {
...
+        const hookFn = testRunHookDefinition.code
+        if (!hookFn.name.startsWith('wdioHook')) {
+            const userHookBeforeTestRun = (...args) => hookFn(args)
+            testRunHookDefinition.code = userHookBeforeTestRun
+        }
+    }))
+}
+
+export function wrapStepWithHooks (code, config) {
+    const userStepFn = async (...args) => {
```
Comment (mgrybyk) [inline packages/wdio-cucumber-framework/src/utils.js]: As far as this one is async function I have to either await the result in executeSync or wrap it with future in a similar way to wrapCommand in wdio/sync, but I don't really like the last option.

**Comment by mgrybyk** [comment]
Added smoke  tests for cucumber to ensure hooks and steps are running sync, will add unit tests later on

**Comment by mgrybyk** [comment]
work completed

**Comment by christian-bromann** [comment]
Thanks so much, I will take a closer look on Monday.

**Comment by mgrybyk** [comment]
@christian-bromann thank you!
Added one very last commit, noticed that step before/after hooks error are not printed out.

**Comment by mgrybyk** [comment]
resolved all the conflicts and made changes according to https://github.com/webdriverio/webdriverio/pull/4308

**Comment by miguelyoobic95** [comment]
Looking forward to seeing this merged, we will test the new version straight away once it is :) thanks a lot guys

File: packages/wdio-cucumber-framework/src/index.js [hunk 4]
```diff
@@ -33,14 +31,23 @@ class CucumberAdapter {
...
+            /**
+             * wdio hooks should be added before spec files are loaded
+             */
+            this.addWdioHooks(this.config)
             this.loadSpecFiles()
-            this.wrapSteps()
+            this.wrapSteps(this.config)
+
+            /**
+             * we need to somehow indentify is function is step or hook
```
Comment (erwinheitzman) [inline packages/wdio-cucumber-framework/src/index.js]: identify

File: packages/wdio-cucumber-framework/src/utils.js [hunk 5]
```diff
@@ -92,15 +94,21 @@ export function getUniqueIdentifier (target, sourceLocation) {
...
-    let message = {
-        ...payload,
+    let content = { ...payload }
+
+    /**
+     * need to convert Error to plain object, otherwise it is lost on process.send
+     */
+    if (payload.error && payload.error.message) {
+        const { name, message, stack } = payload.error
+        content.error = { foo: 'bar', name, message, stack }
```
Comment (erwinheitzman) [inline packages/wdio-cucumber-framework/src/utils.js]: forgot to remove `foo: 'bar'`

File: packages/wdio-cucumber-framework/src/utils.js [hunk 6]
```diff
@@ -92,15 +94,21 @@ export function getUniqueIdentifier (target, sourceLocation) {
...
  */
 export function formatMessage ({ payload = {} }) {
-    let message = {
-        ...payload,
+    let content = { ...payload }
+
+    /**
+     * need to convert Error to plain object, otherwise it is lost on process.send
+     */
+    if (payload.error && payload.error.message) {
```
Comment (erwinheitzman) [inline packages/wdio-cucumber-framework/src/utils.js]: I have seen situations where an error was thrown but was missing a message, should we take this into account?

File: packages/wdio-cucumber-framework/src/utils.js [hunk 7]
```diff
@@ -160,3 +168,99 @@ export function getStepFromFeature(feature, pickle, stepIndex, sourceLocation) {
 
     return targetStep
 }
+
+/**
+ * @param {object[]} result cucumber global result object
+ */
+export function getDataFromResult (result) {
```
Comment (erwinheitzman) [inline packages/wdio-cucumber-framework/src/utils.js]: I think the following would be a lot cleaner
```
export const getDataFromResult ([ uri, feature, ...scenarios ]) => ({ uri, feature, scenarios })
```

File: packages/wdio-cucumber-framework/src/utils.js [hunk 8]
```diff
@@ -160,3 +168,99 @@ export function getStepFromFeature(feature, pickle, stepIndex, sourceLocation) {
...
+ * @param {object} options `Cucumber.supportCodeLibraryBuilder.options`
+ */
+export function setUserHookNames (options) {
+    const hooks = [
+        'beforeTestRunHookDefinitions',
+        'beforeTestCaseHookDefinitions',
+        'afterTestCaseHookDefinitions',
+        'afterTestRunHookDefinitions',
+    ]
+    hooks.forEach(hookName => options[hookName].forEach(testRunHookDefinition => {
```
Comment (erwinheitzman) [inline packages/wdio-cucumber-framework/src/utils.js]: no returning here so the following is preferred:
```
hooks.forEach(hookName => {
...
})
```

File: packages/wdio-cucumber-framework/src/utils.js [hunk 9]
```diff
@@ -160,3 +168,99 @@ export function getStepFromFeature(feature, pickle, stepIndex, sourceLocation) {
...
+ * notify `WDIOCLInterface` about failure in hook
+ * we need to do it this way because `beforeStep` and `afterStep` are not real hooks.
+ * Otherwise hooks failure are lost.
+ * @param {Array}   hookResults hook functions results array
+ * @param {string}  cid         cid
+ */
+export function notifyStepHookError (hookResults = [], cid) {
+    hookResults
+        .filter(result => result && result instanceof Error)
+        .some(r => {
```
Comment (erwinheitzman) [inline packages/wdio-cucumber-framework/src/utils.js]: Array.prototype.some expects a return value to check, see: https://developer.mozilla.org/nl/docs/Web/JavaScript/Reference/Global_Objects/Array/some

File: packages/wdio-cucumber-framework/src/utils.js (hunk 5 above)
Comment (mgrybyk) [reply packages/wdio-cucumber-framework/src/utils.js]: lol

File: packages/wdio-cucumber-framework/src/utils.js (hunk 7 above)
Comment (mgrybyk) [reply packages/wdio-cucumber-framework/src/utils.js]: result object has no such properties

File: packages/wdio-cucumber-framework/src/utils.js [hunk 10]
```diff
@@ -160,3 +168,96 @@ export function getStepFromFeature(feature, pickle, stepIndex, sourceLocation) {
...
+/**
+ * notify `WDIOCLInterface` about failure in hook
+ * we need to do it this way because `beforeStep` and `afterStep` are not real hooks.
+ * Otherwise hooks failure are lost.
+ * @param {string}  hookName    name of the hook
+ * @param {Array}   hookResults hook functions results array
+ * @param {string}  cid         cid
+ */
+export function notifyStepHookError (hookName, hookResults = [], cid) {
+    const result = hookResults.filter(result => result && result instanceof Error)[0]
```
Comment (erwinheitzman) [inline packages/wdio-cucumber-framework/src/utils.js]: `hookResults.find(result => result instanceof Error)` should do the trick (undefined instanceof Error will simply return false and find returns the first match)

**Comment by mgrybyk** [comment]
It would be good to merge this one before https://github.com/webdriverio/webdriverio/pull/4354 as far as some changes relates on each other

File: packages/wdio-cucumber-framework/src/utils.js [hunk 11]
```diff
@@ -160,3 +168,96 @@ export function getStepFromFeature(feature, pickle, stepIndex, sourceLocation) {
...
+ */
+export const getDataFromResult = ([{ uri }, feature, ...scenarios]) => ({ uri, feature, scenarios })
+
+/**
+ * wrap every user defined hook with function named `userHookFn`
+ * to identify later on is function a step, user hook or wdio hook.
+ * @param {object} options `Cucumber.supportCodeLibraryBuilder.options`
+ */
+export function setUserHookNames (options) {
+    const hooks = [
```
Comment (christian-bromann) [inline packages/wdio-cucumber-framework/src/utils.js]: We can move this to the constants.js file or to the top of the file

File: packages/wdio-cucumber-framework/src/utils.js [hunk 12]
```diff
@@ -160,3 +168,96 @@ export function getStepFromFeature(feature, pickle, stepIndex, sourceLocation) {
...
+        'beforeTestRunHookDefinitions',
+        'beforeTestCaseHookDefinitions',
+        'afterTestCaseHookDefinitions',
+        'afterTestRunHookDefinitions',
+    ]
+    hooks.forEach(hookName => {
```
Comment (christian-bromann) [inline packages/wdio-cucumber-framework/src/utils.js]:
```suggestion
    hooks.forEach(hookName => options[hookName].forEach(testRunHookDefinition => {
```

Comment (mgrybyk) [reply packages/wdio-cucumber-framework/src/utils.js]: @christian-bromann isn't it conflicting with what @erwinheitzman suggested? 
See https://github.com/webdriverio/webdriverio/pull/4288#discussion_r312802136

I might be missing something

Comment (christian-bromann) [reply packages/wdio-cucumber-framework/src/utils.js]: yeah, it is a code style thing where we don't have any eslint rules enforcing anything. I personally like if we can save indentation if possible and where it makes sense. I don't have a hard opinion on this.

**Comment by christian-bromann** [comment]
any other comments @erwinheitzman ?

**Comment by mgrybyk** [comment]
resolved conflicts, any other comments @christian-bromann @erwinheitzman ?

**Comment by miguelyoobic95** [comment]
Is this supposed to be working ? We have upgraded and I get my first step on my test being called beforeFeature which is causing quite a bit of issues

**Comment by mgrybyk** [comment]
Please raise an issue with proper description.

**Comment by miguelyoobic95** [comment]
Do you want me to raise an issue or would this be related to this one ? Basically we have a `beforeFeature` hook with some async code and the first `Given` of the cucumber test is still triggered before all the async actions in `beforeFeature` have finished running. Would `beforeAll` and `afterAll` be better replacements ?

**Comment by mgrybyk** [comment]
It's not related to the PR, I don't know anything about your setup, looks like some configuration issue.
If you still have some problems please raise an issue with proper description and steps to reproduce. 
Thank you

**Comment by miguelyoobic95** [comment]
Hey guys, I still have issues, will file a bug report with a repro case as soon as possible. 

Is there an example repo where running functions inside the hooks is working ? Would like to have a look at one so I can check whether our setup is correct. 

Weirdly the code within the hooks on our side only seems to be able to run a console log and it stops running correctly as soon as I add anything else with it stops running, almost like it terminates completely. Any help with examples would be great, I will link the issue here once I raise it :)

**Pull request merged**